torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -967,12 +967,37 @@ void InitXlaModuleBindings(py::module m) {
         [](const at::Tensor& tensor) { return GetTensorViewAliasId(tensor); });
   m.def("_xla_get_tensor_id",
         [](const at::Tensor& tensor) { return GetTensorId(tensor); });
-  m.def("_xla_get_devices",
-        []() { return xla::GetComputationClient()->GetLocalDevices(); });
-  m.def("_xla_num_devices",
-        []() { return xla::GetComputationClient()->GetNumDevices(); });
-  m.def("_xla_get_all_devices",
-        []() { return xla::GetComputationClient()->GetAllDevices(); });
+  m.def("_xla_get_devices", []() {
+    if (ShardingUtil::UseVirtualDevice()) {
+      // Under SPMD context, there is only one virtual devices from user
+      // perspective.
+      std::vector<std::string> all_devices =
+          xla::GetComputationClient()->GetAllDevices();
+      all_devices.resize(1);
+      return all_devices;
+    } else {
+      return xla::GetComputationClient()->GetLocalDevices();
+    }
+  });
+  m.def("_xla_num_devices", []() -> int64_t {
+    if (ShardingUtil::UseVirtualDevice()) {
+      return 1;
+    } else {
+      return xla::GetComputationClient()->GetNumDevices();
+    }
+  });
+  m.def("_xla_get_all_devices", []() {
```

Comment:
This would cause `global_device_count()` to return 1, but we rely on that API to construct the mesh

Reply:
yea, I am thinking about has 2 concepts, `num_devices` -> number of xla devices, in SPMD case it will be 1. and another concept is `num_real_devices` which will return the actual TPU devices. I am still debating whether it make things more clear or more confusing..